### PR TITLE
snap 2.46 (new formula)

### PR DESCRIPTION
**Pull request by sergiusens**
Signed-off-by: Sergio Schvezov <sergio.schvezov@canonical.com>

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

File: Formula/snap.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  head do
+    url "https://github.com/snapcore/snapd.git"
+  end
+
+  depends_on "go" => :build
+  depends_on "squashfs"
+
+  def install
+    ENV["GOPATH"] = buildpath
+    ENV["GOCACHE"] = "#{HOMEBREW_CACHE}/go_cache"
```
Comment (dawidd6) [inline Formula/snap.rb:21]: What is this for?

File: Formula/snap.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      if version.head?
+        system "./get-deps.sh"
+        system "./mkversion.sh"
+      elsif revision.positive?
+        system "./mkversion.sh", "#{version}-#{revision}"
+      else
+        system "./mkversion.sh", version
+      end
+
+      system "go", "build", "-o", bin/"snap", "./cmd/snap"
```
Comment (dawidd6) [inline Formula/snap.rb:34]:
```suggestion
      system "go", "build", *std_go_args, "./cmd/snap"
```

File: Formula/snap.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      bash_completion.install "data/completion/bash/snap"
+
+      # Build zsh completion.
+      zsh_completion.install "data/completion/zsh/_snap"
+
+      # Build manpage.
+      system "sh", "-c", "#{bin}/snap help --man > snap.8"
+      man8.install "snap.8"
+
+      prefix.install_metafiles
```
Comment (dawidd6) [inline Formula/snap.rb:46]:
```suggestion
```

File: Formula/snap.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    end
+  end
+
+  test do
+    (testpath/"pkg/meta").mkpath
+    (testpath/"pkg/meta/snap.yaml").write "name: test-snap\nversion: 1.0.0\nsummary: s\ndescription: d"
```
Comment (dawidd6) [inline Formula/snap.rb:52]:
```suggestion
    (testpath/"pkg/meta/snap.yaml").write <<~EOS
      name: test-snap
      version: 1.0.0
      summary: s
      description: d
    EOS
```

File: Formula/snap.rb (hunk 1 above)
Comment (sergiusens) [reply Formula/snap.rb:21]: This is for faster rebuilds, I guess it makes sense locally, can be removed if desired

Comment (dawidd6) [reply Formula/snap.rb:21]: We already set it in `brew`. Yea, remove it please.

File: Formula/snap.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  stable do
+    version "2.46"
+    url "https://github.com/snapcore/snapd/releases/download/#{version}/snapd_#{version}.vendor.tar.xz"
+    sha256 "c4f532018ca9d2a5f87a95909b3674f8e299e97ba5cb5575895bcdd29be23db3"
+  end
+
+  head do
+    url "https://github.com/snapcore/snapd.git"
+  end
```
Comment (bayandin) [inline Formula/snap.rb:14]:
```suggestion
  desc "The snap tool enable systems to work with .snap files"
  homepage "https://snapcraft.io/"
  url "https://github.com/snapcore/snapd/releases/download/2.46/snapd_2.46.vendor.tar.xz"
  sha256 "c4f532018ca9d2a5f87a95909b3674f8e299e97ba5cb5575895bcdd29be23db3"
  license "GPL-3.0-only"
  head "https://github.com/snapcore/snapd.git"
```

File: Formula/snap.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    ENV["GOPATH"] = buildpath
+    ENV["GOCACHE"] = "#{HOMEBREW_CACHE}/go_cache"
+    (buildpath/"src/github.com/snapcore/snapd").install buildpath.children
+
+    cd "src/github.com/snapcore/snapd" do
+      if version.head?
+        system "./get-deps.sh"
+        system "./mkversion.sh"
+      elsif revision.positive?
+        system "./mkversion.sh", "#{version}-#{revision}"
```
Comment (bayandin) [inline Formula/snap.rb:29]:
```suggestion
```

File: Formula/snap.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+      system "go", "build", "-o", bin/"snap", "./cmd/snap"
+
+      # Build bash completion.
+      bash_completion.install "data/completion/bash/snap"
+
+      # Build zsh completion.
+      zsh_completion.install "data/completion/zsh/_snap"
+
+      # Build manpage.
```
Comment (bayandin) [inline Formula/snap.rb:42]:
```suggestion
      bash_completion.install "data/completion/bash/snap"
      zsh_completion.install "data/completion/zsh/_snap"

```

**Review by bayandin — CHANGES_REQUESTED** [review]

**Comment by sergiusens** [comment]
Thanks for the review @dawidd6, I believe I addressed your requests

**Comment by sergiusens** [comment]
Thanks for the review @bayandin, I believe I addressed your requests as well.

File: Formula/snap.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  url "https://github.com/snapcore/snapd/releases/download/2.46/snapd_2.46.vendor.tar.xz"
+  sha256 "c4f532018ca9d2a5f87a95909b3674f8e299e97ba5cb5575895bcdd29be23db3"
+  license "GPL-3.0-only"
+  head "https://github.com/snapcore/snapd.git"
+
+  depends_on "go" => :build
+  depends_on "squashfs"
+
+  def install
+    ENV["GOPATH"] = buildpath
```
Comment (SMillerDev) [inline Formula/snap.rb:13]: Can this work without setting GOPATH?

File: Formula/snap.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    (buildpath/"src/github.com/snapcore/snapd").install buildpath.children
+
+    cd "src/github.com/snapcore/snapd" do
+      if version.head?
+        system "./get-deps.sh"
```
Comment (SMillerDev) [inline Formula/snap.rb:18]: What does this download?

Comment (anonymouse64) [reply Formula/snap.rb:18]: this essentially just installs govendor and the deps declared in vendor.json using govendor, you could call govendor directly if one was so inclined

Comment (SMillerDev) [reply Formula/snap.rb:18]: Any downloads needs to be resource blocks or using a check-summed system.

Comment (anonymouse64) [reply Formula/snap.rb:18]: I'll let @sergiusens sort out how to do that in this recipe, just wanted to clarify what that script does as a snapd developer myself. Note also that govendor does checksum things when it downloads them, the vendor.json contains the hashes: https://github.com/snapcore/snapd/blob/master/vendor/vendor.json

Comment (dawidd6) [reply Formula/snap.rb:18]: @anonymouse64 any reason why you don't switch to Go Modules?

Comment (anonymouse64) [reply Formula/snap.rb:18]: We have to support very old Linux distros, we want to switch but can't yet until old distros have backported newer go versions

File: Formula/snap.rb (hunk 8 above)
Comment (dawidd6) [reply Formula/snap.rb:13]: It cannot. See @anonymouse64 answer below.

File: Formula/snap.rb (hunk 9 above)
Comment (sergiusens) [reply Formula/snap.rb:18]: To clarify, this is only required for `--HEAD`, which I can remove from the formula if this becomes a blocker. The "released path"  is using vendored assets.

Comment (SMillerDev) [reply Formula/snap.rb:18]: I'd say that's the better solution here.

Comment (sergiusens) [reply Formula/snap.rb]: I have removed head and with it the call to get-deps.sh

**Comment by sergiusens** [comment]
db88e0e: I removed the option for --HEAD and added a specific "version" as the version was being incorrectly detected (2.46.vendor)

File: Formula/snap.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    (buildpath/"src/github.com/snapcore/snapd").install buildpath.children
+
+    cd "src/github.com/snapcore/snapd" do
+      system "./mkversion.sh", version
+      system "go", "build", *std_go_args, "./cmd/snap"
+
+      bash_completion.install "data/completion/bash/snap"
+      zsh_completion.install "data/completion/zsh/_snap"
+
+      system "sh", "-c", "#{bin}/snap help --man > snap.8"
```
Comment (SMillerDev) [inline Formula/snap.rb:23]: This will be cleaner
```suggestion
     (man8"/snap.8").write Utils.safe_popen_read("#{bin}/snap", "help", "--man")
```

Comment (sergiusens) [reply Formula/snap.rb]: Fixed in ed6010b

**Review by SMillerDev — APPROVED** [review]

**Pull request closed without merging**